### PR TITLE
add webrtc demo

### PR DESCRIPTION
**Pull request by pretentious7**

### CODE REVIEW

File: js-pkg/apps/demos/src/pages/index.tsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/index.tsx:34]: Since this actually shows up on the site, let's capitalize and add a period.

File: js-pkg/apps/demos/src/components/chat.jsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/chat.jsx:20]: Not for this PR, but it would be nice to interleave this down the road

File: js-pkg/apps/demos/src/components/driftdbutils.jsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/driftdbutils.jsx:2]: From offline convo: maybe call this `signaling-utils.ts`?

File: js-pkg/apps/demos/src/components/driftdbutils.jsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/driftdbutils.jsx:1]: From offline convo: maybe call this file `signaling-utils.ts`?

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:49]: Should this be `new conn` or `new connection` or something like that?

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:77]: From offline convo: these `setIsReady` don't seem to be defined anywhere

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:126]: You should throw an actual error here instead of just a string: `throw new Error('unknown message')`

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:113]: Should this be a `continue` instead of a `break`?

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:38]: from offline convo: you could get rid of the `theyAreReady` state above and just rely on the `theirDataChannelRef`

```js
(msg) => {
  if (theirdataChannelRef.current) theirdataChannelRef.current.send(msg)
  else console.log('unsent:', msg)
}
```

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:11]: From offline convo, maybe you could pass this list of functions to `useWebRTCConnection()`?

File: js-pkg/apps/demos/src/components/chat.jsx
Comment (paulgb) [inline js-pkg/apps/demos/src/components/chat.jsx:40]: I'd prefer `{' send '}` if this is what you really want, but I think we probably want to use tailwind to create the padding instead.

**Comment by pretentious7** [comment]
Screenshot from 2023-03-13 17-03-12

tada, this is ready now!

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:150]: nit: right now this function is expected to either return a number, NaN, undefined, or throw an Error. Might be simpler to just have this return either a number or null?

I think that would just mean dropping the `throw new Error` block, changing the st.currentRoundTripTime check to `Number.isFinite(st.currentRoundTripTime)` and then adding a `return null` at the end.

Should simplify some of the calling code, too.

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:30]: See other comment about simplifying the expected behavior of `getLatency`, that way you can remove this try/catch.

Also, technically, we should probably cancel this timeout on unmount, otherwise we could end up calling `setLatency()` on a dead component, but since this is just a demo, maybe it's fine.

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:26]: nit: why the underscore prefix?

File: js-pkg/apps/demos/src/pages/webrtc-demo.jsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/webrtc-demo.jsx:11]: Let's format this so we don't have `75.99999999999999999999998 ms` in the table.

Something like this maybe?:

```js
function formatFloat(num, decimals) {
  const mult = Math.pow(10, decimals)
  return Math.round(num * mult) / mult
}
```

File: js-pkg/apps/demos/src/components/chat.jsx
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/chat.jsx:42]: Looks like there are some cases where latency doesn't get reported. When that happens, it might be best to just not show any latency in the UI at all because it makes the demo look broken.

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js:147]: ye done. thank you!

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:26]: You'll need to assign this to `latency_setter`. (Maybe `timeoutId` is a better name btw?)

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [reply js-pkg/apps/demos/src/components/webrtc.js:147]: Looks like this can still return `null` without a return at the end of the fn?

Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js:151]: ah that returns undefined not null
my bad

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js:31]: done :+1:

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js]: aaaaa, always forget that. thank you.

File: js-pkg/apps/demos/src/components/webrtc.js
Comment (rolyatmax) [inline js-pkg/apps/demos/src/components/webrtc.js:143]: I'm confused by the comment - `getStats()` returns `undefined` before a connection is established? Why is the `try/catch` necessary?

Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js:143]: the optional accessor throws an error

Comment (pretentious7) [reply js-pkg/apps/demos/src/components/webrtc.js:143]: but I guess this is unwieldy, better to just check that's defined.

File: js-pkg/apps/demos/next.config.js
Comment (paulgb) [inline js-pkg/apps/demos/next.config.js:3]: #77 will make this unnecessary

**Review by paulgb — COMMENTED** [review]
LGTM, but @rolyatmax was deeper into this review so I'll let him make the final call. Left one note; if we merge this after #77 we can avoid turning off safe mode.

**Review by rolyatmax — APPROVED** [review]
Looks good to me!

**Pull request merged**
